### PR TITLE
Docsp 7000: Create productionDeployJob

### PR DESCRIPTION
**Pull request by madelinezec**
Plumbing complete. Ready for more intensive testing.

### CODE REVIEW

File: worker/jobTypes/productionDeployJob.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+      return false;
+    }
+  }
+  
+  async function verifyBranchConfiguredForPublish(currentJob) {
+    const repoObject = { repoOwner: currentJob.payload.repoOwner, repoName: currentJob.payload.repoName};
+    const repoContent = await workerUtils.getRepoPublishedBranches(repoObject);
+    const publishedBranches = repoContent['content']['git']['branches']['published']
+
+    if (publishedBranches.includes(currentJob.payload.branchName)) {
```
Comment (skerschb) [inline worker/jobTypes/productionDeployJob.js]: style point -- you could simply return publishedBranches.includes(currentJob.payload.branchName);

File: worker/jobTypes/productionDeployJob.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+  let branchext = '';
+  let isMaster = true;
+
+  if (currentJob.payload.branchName !== 'master') {
+    branchext = '-' + currentJob.payload.branchName;
+    isMaster = false;
+  }
+
+  if (isMaster) {
```
Comment (skerschb) [inline worker/jobTypes/productionDeployJob.js]: we can get rid of any logic around checking for master -- everything moving through this job type is production-bound and as long as the user is entitled to publish it, can go to the same publish calls in S3.

File: worker/jobTypes/githubJob.js [hunk 3]
```diff
@@ -42,13 +42,20 @@ class GitHubJobClass {
...
+    }
+    return true;
+  }
 
   // cleanup before pulling repo
   async cleanup(logger) {
     const currentJob = this.currentJob;
     logger.save(`${'(rm)'.padEnd(15)}Cleaning up repository`);
     try {
-      workerUtils.removeDirectory(`repos/${this.getRepoDirName(currentJob)}`);
```
Comment (skerschb) [inline worker/jobTypes/githubJob.js]: why did we swap this out for a duplicated method?

File: worker/jobTypes/githubJob.js [hunk 4]
```diff
@@ -59,6 +66,20 @@ class GitHubJobClass {
     });
   }
 
+    // Function that rejects function (promise) after (seconds) seconds with error (errMsg)
+    promiseTimeoutS(seconds, promise, errMsg) {
```
Comment (skerschb) [inline worker/jobTypes/githubJob.js]: same here?

File: worker/jobTypes/productionDeployJob.js [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    safeString(currentJob.payload.repoName) &&
+    safeString(currentJob.payload.repoOwner) &&
+    safeString(currentJob.payload.branchName)
+  ) {
+    return true;
+  }
+  throw invalidJobDef;
+}
+  // Function that rejects function (promise) after (seconds) seconds with error (errMsg)
+  function promiseTimeoutS(seconds, promise, errMsg) {
```
Comment (skerschb) [inline worker/jobTypes/productionDeployJob.js]: another dup from utils.

File: worker/jobTypes/productionDeployJob.js [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+async function runGithubProdPush(currentJob) {
+  const ispublishable = await verifyBranchConfiguredForPublish(currentJob);
+  const userIsEntitled = await verifyUserEntitlements(currentJob);
+
+  if (ispublishable === false) {
+    workerUtils.logInMongo(
+      currentJob,
+      `${"(BUILD)".padEnd(15)} You are trying to 
+    
+    to production a branch that is not configured for publishing`
```
Comment (skerschb) [inline worker/jobTypes/productionDeployJob.js]: looks like a weird merge issue

File: worker/jobTypes/productionDeployJob.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    );
+    throw new Error("entitlement failed");
+    process.exit;
```
Comment (skerschb) [inline worker/jobTypes/productionDeployJob.js]: is this a fatal error?

File: worker/jobTypes/productionDeployJob.js [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+  }
+  if (userIsEntitled === false) {
+    workerUtils.logInMongo(
+      currentJob,
+      `${"(BUILD)".padEnd(
+        15
+      )} failed, you are not entitled to build or deploy (${currentJob.payload.repoOwner}/${currentJob.payload.repoName}) for master branch`
+    );
+    throw new Error("entitlement failed");
+    process.exit;
```
Comment (skerschb) [inline worker/jobTypes/productionDeployJob.js]: same here -- I don't think we want to run process.exit here.

File: worker/jobTypes/S3Publish.js [hunk 9]
```diff
@@ -32,17 +59,74 @@ class S3PublishClass {
...
-    // todo
-  }
+  async pushToProduction(logger) {    
+    const deployCommands = [
+      `. /venv/bin/activate`,
+      `cd repos/${this.GitHubJob.getRepoDirName(this.GitHubJob.currentJob)}`,
+      `make stage`
+    ];
 
+    // the way we now build is to search for a specific function string in worker.sh
```
[the remark this answers is not in the document]
Comment (danielborowski) [reply worker/jobTypes/S3Publish.js]: agreed

File: worker/jobTypes/productionDeployJob.js (hunk 1 above)
Comment (danielborowski) [reply worker/jobTypes/productionDeployJob.js]: +1

File: worker/jobTypes/productionDeployJob.js (hunk 7 above)
Comment (danielborowski) [reply worker/jobTypes/productionDeployJob.js]: removed this line

File: worker/jobTypes/productionDeployJob.js (hunk 8 above)
Comment (danielborowski) [reply worker/jobTypes/productionDeployJob.js]: removed this line

**Comment by skerschb** [comment]
let's merge this to stage for testing against the staging environment. Can we coordinate that with the change to the deploy slack app? I believe it's best to kick the tires on this before we start testing the commit work.

**Pull request merged**